src/components/files/createThumbnails.ts

```diff
@@ -0,0 +1,77 @@
+// Maps file types (mimetype or ext) to generic
+// thumbnail files
+const genericThumbnails: { [key: string]: string } = {
+  image: 'noun_Image_1570203.png',
+  doc: 'noun_word_3515287.png',
+  docx: 'noun_word_3515287.png',
+  pdf: 'noun_PDF_3283219.png',
+  text: 'noun_word_3515287.png',
+  msword: 'noun_word_3515287.png',
+  file: 'noun_File_3764922.png',
+}
+
+import * as config from '../../config.json'
+import path from 'path'
+import fs from 'fs'
+const sharp = require('sharp')
+const { thumbnailMaxWidth, thumbnailMaxHeight } = config
+
+interface ThumbnailInput {
+  filesPath: string
+  unique_id: string
+  basename: string
+  ext: string
+  mimetype: string
+}
+
+// We use a combination of mimetype (type/subtype) and file extension to
+// determine which conversion process to use, or which generic
+// thumbnail to return
+const createThumbnail = async ({
+  filesPath,
+  unique_id,
+  basename,
+  ext,
+  mimetype,
+}: ThumbnailInput) => {
+  const { type, subtype } = splitMimetype(mimetype)
+
+  const origFilePath = path.join(filesPath, `${basename}_${unique_id}${ext}`)
+  const thumbnailFilePath = path.join(filesPath, `${basename}_${unique_id}_thumb`) // No ext, added after conversion
+
+  if (type === 'image') {
+    try {
+      const { format } = await sharp(origFilePath)
+        .resize(thumbnailMaxWidth, thumbnailMaxHeight, {
+          fit: 'inside',
+          withoutEnlargement: true,
+        })
+        .toFile(thumbnailFilePath)
+      await fs.rename(thumbnailFilePath, `${thumbnailFilePath}.${format}`, () => {})
+      return `${basename}_${unique_id}_thumb.${format}`
+    } catch {
+      return getGenericThumbnailPath(type)
+    }
+  } else if (mimetype === 'application/pdf') {
+    // TO-DO Generate PDF previews/thumbnails
+    return getGenericThumbnailPath('pdf')
+  } else if (['.pdf', '.doc', '.docx'].includes(ext))
```

Comment:
Yeah, it's not ideal, but the problem is there is not a lot of consistency. For example, all "image" file types have "image" in the first part of their mimetype, so it makes a lot more sense to match by type rather than think of all possible image extensions. And on the other hand, there's no generic mimetype that will capture all the possible types of file that should get the generic "DOC" thumbnail, so you need to really match by extension. I think once we have a clearer idea of what our final set of generic thumbnails will be and what the possible types and extensions are, we can make a simpler, definitive mapping. Current version was just to make sure it works for a few basic, common format we're likely to use.